commons/src/main/java/org/infinispan/commons/marshall/MarshallUtil.java

```diff
@@ -21,27 +22,327 @@
 
    private static final byte NULL_VALUE = -1;
 
-   public static void marshallCollection(Collection<?> c, ObjectOutput out) throws IOException {
-      UnsignedNumeric.writeUnsignedInt(out, c.size());
-      for (Object o : c) {
-         out.writeObject(o);
+   /**
+    * Marshall the {@code map} to the {@code ObjectOutput}.
+    * <p>
+    * {@code null} maps are supported.
+    *
+    * @param map {@link Map} to marshall.
+    * @param out {@link ObjectOutput} to write. It must be non-null.
+    * @param <K> Key type of the map.
+    * @param <V> Value type of the map.
+    * @param <T> Type of the {@link Map}.
+    * @throws IOException If any of the usual Input/Output related exceptions occur.
+    */
+   public static <K, V, T extends Map<K, V>> void marshallMap(T map, ObjectOutput out) throws IOException {
+      final int mapSize = map == null ? NULL_VALUE : map.size();
+      marshallInt(out, mapSize);
+      if (mapSize <= 0) return;
+
+      for (Map.Entry<K, V> me : map.entrySet()) {
+         out.writeObject(me.getKey());
+         out.writeObject(me.getValue());
       }
    }
 
-   public static void marshallMap(Map<?, ?> map, ObjectOutput out) throws IOException {
-      int mapSize = map.size();
-      UnsignedNumeric.writeUnsignedInt(out, mapSize);
-      if (mapSize == 0) return;
+   /**
+    * Unmarshall the {@link Map}.
+    * <p>
+    * If the marshalled map is {@link null}, then the {@link MapBuilder} is not invoked.
+    *
+    * @param in      {@link ObjectInput} to read.
+    * @param builder {@link MapBuilder} to create the concrete {@link Map} implementation.
+    * @return The populated {@link Map} created by the {@link MapBuilder} or {@code null}.
+    * @throws IOException            If any of the usual Input/Output related exceptions occur.
+    * @throws ClassNotFoundException If the class of a serialized object cannot be found.
+    * @see {@link #marshallMap(Map, ObjectOutput)}
+    */
+   public static <K, V, T extends Map<K, V>> T unmarshallMap(ObjectInput in, MapBuilder<K, V, T> builder) throws IOException, ClassNotFoundException {
+      final int size = unmarshallInt(in);
+      if (size == NULL_VALUE) {
+         return null;
+      }
+      final T map = Objects.requireNonNull(builder, "MapBuilder must be non-null").build();
```

Comment:
Since we know the size the map will have, why not enhance the MapBuilder to also support an initial size to avoid resizing while unmarshalling.